### DISCUSSION TITLE
How to use forwardRef component with MDXProvider and keep TypeScript happy?

### DISCUSSION THREAD

**Discussion by igordanchenko** [mdx-js/mdx#2086] [Q&A] [answered]
I'm having a difficult time making TypeScript happy while using a `React.forwardRef`-wrapped component with `MDXProvider`.

Here is a minimal example that does not type-check with TypeScript.

```jsx
// _app.tsx

const components = {
    a: Link,
};

...

<MDXProvider components={components}>
    ...
</MDXProvider>
```

```jsx
// Link.tsx

type LinkProps = React.AnchorHTMLAttributes<HTMLAnchorElement>;

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(({ children, ...rest }, ref) => (
    <a ref={ref} {...rest}>
        {children}
    </a>
));
Link.displayName = "Link";
```

TypeScript throws the following error:

```
Type error: Type '{ h1: FC<{ children?: ReactNode; }>; h2: FC<{ children?: ReactNode; }>; a: ForwardRefExoticComponent<LinkProps & RefAttributes<HTMLAnchorElement>>; }' is not assignable to type 'MDXComponents'.
  Type '{ h1: FC<{ children?: ReactNode; }>; h2: FC<{ children?: ReactNode; }>; a: ForwardRefExoticComponent<LinkProps & RefAttributes<HTMLAnchorElement>>; }' is not assignable to type '{ a?: keyof IntrinsicElements | Component<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>> | undefined; ... 173 more ...; view?: keyof IntrinsicElements | ... 1 more ... | undefined; }'.
    Types of property 'a' are incompatible.
      Type 'ForwardRefExoticComponent<LinkProps & RefAttributes<HTMLAnchorElement>>' is not assignable to type 'keyof IntrinsicElements | Component<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>> | undefined'.
        Type 'ForwardRefExoticComponent<LinkProps & RefAttributes<HTMLAnchorElement>>' is not assignable to type 'FunctionComponent<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>>'.
          Types of parameters 'props' and 'props' are incompatible.
            Type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>' is not assignable to type 'LinkProps & RefAttributes<HTMLAnchorElement>'.
              Type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>' is not assignable to type 'RefAttributes<HTMLAnchorElement>'.
                Types of property 'ref' are incompatible.
                  Type 'LegacyRef<HTMLAnchorElement> | undefined' is not assignable to type 'Ref<HTMLAnchorElement> | undefined'.
                    Type 'string' is not assignable to type 'Ref<HTMLAnchorElement> | undefined'.
```

I found 2 workarounds so far but both aren't pretty.

1) Create an extra wrapper and provide that wrapper component to MDXProvider:

```jsx
export const LinkWithoutRef: React.FC<React.PropsWithoutRef<Parameters<typeof Link>[0]>> = ({ children, ...rest }) => (
    <Link {...rest}>{children}</Link>
);
```

2) @ts-ignore

Am I missing some obvious solution here? I'd appreciate expert advice.

**Comment by wooorm** [comment, marked as answer]
To get types to work, it is important to use the same types on your end as are allowed in the MDX types.
The MDX types are here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mdx/types.d.ts#L22, which expands to `JSX.IntrinsicElements['a']`.

**Reply by igordanchenko** [reply]
Yes, that's effectively the type I was using. I'll update for clarity:

```jsx
type LinkProps = JSX.IntrinsicElements['a'];

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(({children, ...rest}, ref) => (
    <a ref={ref} {...rest}>{children}</a>
));
Link.displayName = "Link";
```

This results in the same error:

```
Type error: Type '{ a: ForwardRefExoticComponent<Pick<LinkProps, "key" | keyof AnchorHTMLAttributes<HTMLAnchorElement>> & RefAttributes<HTMLAnchorElement>>; }' is not assignable to type 'MDXComponents'.
  Type '{ a: ForwardRefExoticComponent<Pick<LinkProps, "key" | keyof AnchorHTMLAttributes<HTMLAnchorElement>> & RefAttributes<HTMLAnchorElement>>; }' is not assignable to type '{ a?: keyof IntrinsicElements | Component<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>> | undefined; ... 173 more ...; view?: keyof IntrinsicElements | ... 1 more ... | undefined; }'.
    Types of property 'a' are incompatible.
      Type 'ForwardRefExoticComponent<Pick<LinkProps, "key" | keyof AnchorHTMLAttributes<HTMLAnchorElement>> & RefAttributes<HTMLAnchorElement>>' is not assignable to type 'keyof IntrinsicElements | Component<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>> | undefined'.
        Type 'ForwardRefExoticComponent<Pick<LinkProps, "key" | keyof AnchorHTMLAttributes<HTMLAnchorElement>> & RefAttributes<HTMLAnchorElement>>' is not assignable to type 'FunctionComponent<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>>'.
          Types of parameters 'props' and 'props' are incompatible.
            Type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>' is not assignable to type 'Pick<LinkProps, "key" | keyof AnchorHTMLAttributes<HTMLAnchorElement>> & RefAttributes<HTMLAnchorElement>'.
              Type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>' is not assignable to type 'RefAttributes<HTMLAnchorElement>'.
                Types of property 'ref' are incompatible.
                  Type 'LegacyRef<HTMLAnchorElement> | undefined' is not assignable to type 'Ref<HTMLAnchorElement> | undefined'.
                    Type 'string' is not assignable to type 'Ref<HTMLAnchorElement> | undefined'.
```

**Reply by wooorm** [reply]
I don’t know 🤷‍♂️
/cc @remcohaszing maybe?

**Reply by igordanchenko** [reply]
It feels like there may be a bug in React types, but I can't make a strong argument since legacy refs predate my React experience. `LegacyRef` type is coming from `React.ClassAttributes`, so effectively, `@types/mdx` is telling TypeScript "I'm expecting to receive a component here, and if it accepts `ref` attribute, it must be of `LegacyRef` type." But as far as I can tell `@mdx-js` isn't using refs, so it shouldn't care what type of refs the mdx component actually accepts. Omitting the `ref` attribute in the `Component` type definition could solve this TypeScript error, but I'm not sure it's the right way.

```jsx
type Component<Props> = Omit<FunctionComponent<Props> | ClassComponent<Props>, "ref">;
```

**Reply by igordanchenko** [reply]
Huh, here is an interesting case. Adding `ref` attribute to the link in mdx file 

```html
<a href="abc" ref="xyz">abc</a>
```

results in the following runtime React error:

```
react-dom.development.js?ac89:14831 Uncaught Error: Function components cannot have string refs. We recommend using useRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref
    at coerceRef (react-dom.development.js?ac89:14831:1)
    at createChild (react-dom.development.js?ac89:15111:1)
    at reconcileChildrenArray (react-dom.development.js?ac89:15404:1)
    at reconcileChildFibers (react-dom.development.js?ac89:15821:1)
    at reconcileChildren (react-dom.development.js?ac89:19167:1)
    at mountIndeterminateComponent (react-dom.development.js?ac89:20157:1)
    at beginWork (react-dom.development.js?ac89:21587:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js?ac89:4164:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js?ac89:4213:1)
    at invokeGuardedCallback (react-dom.development.js?ac89:4277:1)
    at beginWork$1 (react-dom.development.js?ac89:27451:1)
    at performUnitOfWork (react-dom.development.js?ac89:26557:1)
    at workLoopSync (react-dom.development.js?ac89:26466:1)
    at renderRootSync (react-dom.development.js?ac89:26434:1)
    at recoverFromConcurrentError (react-dom.development.js?ac89:25850:1)
    at performConcurrentWorkOnRoot (react-dom.development.js?ac89:25750:1)
    at workLoop (scheduler.development.js?bcd2:266:1)
    at flushWork (scheduler.development.js?bcd2:239:1)
    at MessagePort.performWorkUntilDeadline (scheduler.development.js?bcd2:533:1)
```

Maybe `@mdx-js` should sanitize attributes and physically omit string `ref`s since they can't be passed down safely?

**Reply by wooorm** [reply]
I don't think that has anything to do with MDX. If you pass a string as ref in normal jsx, you will most likely get the same error.

**Reply by igordanchenko** [reply]
You are correct, same runtime error.

**Reply by remcohaszing** [reply]
I copied the code in the OP into the TypeScript playground and it just works. ([link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDecAsgCIAaACsQG7AAJlihwAvnCIlyAARCCAHgFoAVqgD0lPDFo0YATzBY4AGWAA7ANa8IYdAF5MVGADoAgudwALaAAkAKqwmbjAwUMAARgCuMFioADwBQR7e0ACiADZYIFjmMAB8ANw0NLgQ5qjwZlZwjtjaLgTQAO7IUILYBImBwZ4+UJnZuTAANKYW1sR2+QAUMyzewBmClOZjLhuUleJjlAQAlLX5cDM0cOdw8chwe-ZMexJMGy5bMGL5ZxdfTIvLq2Kfc7xdTID77fbFaqWFyCYCoMAZZD6AByyBytTgACIoZjiiUyhV4GVwOVhg5mICUAAucZWEY0MR4mjxDg8fhCERwYmQcxku7c0l5VDvSnPZnqVk2ATCKAfIA))

I suspect you have multiple versions of `@types/react` installed and this has nothing to do with MDX. This is an issue for any React library, not just `mdx-js/react`. Maybe we should change how  `@mdx-js/react` depends on `@types/react` to avoid these kind of issues. `react` is a peer dependency, so `@types/react` should be too. To not bother people not using TypeScript, we could make it an optional peer dependency. I also wouldn’t mind not specifying the dependency on `@types/react` at all, as I expect anyone to use React and TypeScript to install `@types/react` anyway.

The short-term solution for @igordanchenko is to use [npm overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) or [yarn resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) for now.

**Reply by igordanchenko** [reply]
Unfortunately, I do not see duplicate `@types/react` instances 🤷‍♂️

```
yarn why @types/react
yarn why v1.22.15
[1/4] 🤔  Why do we have the module "@types/react"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "@types/react@18.0.15"
info Has been hoisted to "@types/react"
info Reasons this module exists
   - Specified in "devDependencies"
   - Hoisted from "@types#react-dom#@types#react"
   - Hoisted from "@mdx-js#react#@types#react"
info Disk size without dependencies: "196KB"
info Disk size with unique dependencies: "1.38MB"
info Disk size with transitive dependencies: "1.38MB"
info Number of shared dependencies: 3
✨  Done in 0.24s.
```

I'm using Next.js, and here is how I setup test sandbox:

```sh
npx create-next-app mdxjs-test --typescript
yarn add @next/mdx @mdx-js/react @mdx-js/loader
```

**Reply by igordanchenko** [reply]
@remcohaszing, thank you for looking into this. I'm still unclear why the TS playground doesn't show any errors, but the same error occurs in real-life Next.js and CRA projects. For the time being, I ended up just type-casting to the type MDXProvider expects to receive.  

```jsx
const components = {
    a: Link as React.FC<JSX.IntrinsicElements['a']>
}
```

**Reply by remcohaszing** [reply]
I don’t know why it works in the TypeScript playground, but not in your example. My guess is the playground uses an outdated version of TypeScript.

Anyway, I did manage to reproduce the issue.

MDX uses framework agnostic JSX types. This depends on the `JSX.IntrinsicElements` interface. The problem is that the React types specify `JSX.IntrinsicElements['a']` accepts a `ref` prop of type `LegacyRef | undefined`, but they also specify `React.forwardRef` returns a component that accepts a `ref` prop of type `Ref | undefined`. `LegacyRef` is not assignable to `Ref`.

I think this solution by @igordanchenko is the best one available:

```ts
const components = {
    a: Link as React.FC<JSX.IntrinsicElements['a']>
}
```

We could make the following change to `@types/mdx` to avoid this issue, but that would be a React specific workaround.

```diff
- type FunctionComponent<Props> = (props: Props) => JSX.Element | null;
+ type FunctionComponent<Props> = (props: Omit<Props, 'ref'>) => JSX.Element | null;
```

**Reply by wooorm** [reply]
Isn’t it more of a bug with `@types/react`?

**Comment by schardev** [comment]
> ```ts
> const components = {
>     a: Link as React.FC<JSX.IntrinsicElements['a']>
> }
> ```

The workaround doesn't work now? I got this error:
```
components/mdx-components.tsx:8:3 - error TS2322: Type 'FC<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>>' is not assignable to type 'Component<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>> | undefined'.
  Type 'React.FunctionComponent<React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>>' is not assignable to type 'import("/home/saurabh/workspace/blog-template/node_modules/.pnpm/@types+mdx@2.0.5/node_modules/@types/mdx/types").FunctionComponent<React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>>'.
    Type 'ReactNode' is not assignable to type 'Element | null'.
      Type 'undefined' is not assignable to type 'Element | null'.

8   a: Link as React.FC<JSX.IntrinsicElements["a"]>,
    ~

  node_modules/.pnpm/@types+react@18.2.14/node_modules/@types/react/index.d.ts:3255:13
    3255             a: React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>;
                     ~
    The expected type comes from property 'a' which is declared here on type 'MDXComponents'


```

The only way I got it to ignore is to use MDX's own type definition for a [function component](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1b85e933696518a58ca05d48c31065e0d20430ca/types/mdx/types.d.ts#L4).

**Reply by igordanchenko** [reply]
@schardev, here is my current workaround:

```js
import * as React from "react";
import type { FunctionComponent, MDXComponents } from "mdx/types";

const components: MDXComponents = {
    a: Link as FunctionComponent<React.JSX.IntrinsicElements["a"]>,
};
```

**Reply by schardev** [reply]
This is exactly what I'm using now.

**Reply by remcohaszing** [reply]
I think this is caused by the introduction of `JSX.ElementType` and will be fixed by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65703.

My only note to this workaround is that the `JSX` namespace is an implementation detail you should avoid using. Better use `ComponentPropsWithoutRef`.

```ts
import * as React from "react";
import type { FunctionComponent, MDXComponents } from "mdx/types";

const components: MDXComponents = {
    a: Link as FunctionComponent<React.ComponentPropsWithoutRef<"a">>,
};
```

**Reply by remcohaszing** [reply]
This was fixed in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65703.
